### PR TITLE
Support 4point solid

### PR DESCRIPTION
**Pull request by asmwarrior**
This implement the feature request for the windex target.
https://github.com/JamesBremner/DXF_Viewer/issues/23

### CODE REVIEW

**Comment by asmwarrior** [comment]
In-fact, I have some local changes to the cbp file, but I think you and I are not using the same compiler(the compiler name in cbp file is: you use is `gcc_83`, but I use `gcc`), so I don't pull request those changes.

Another issue is that in my compiler, I have to manually define
`#define TBS_TRANSPARENTBKGND 0x100` in some files
to make the compiler happy.

**Comment by asmwarrior** [comment]
Also, please note that [* add support for drawing the Polygon](https://github.com/JamesBremner/windex/pull/1) is need to build the pull request.

File: src/windex.cpp [hunk 1]
```diff
@@ -256,19 +256,51 @@ void CSolid::Draw( CDxf* dxf )
...
 
+        if (v.size()>0)
+        {
+            for(int i=0;i<v.size()/2;i++)
+            {
+                std::cout << v[2*i] << " " << v[2*i+1] << "\n";
+            }
+            dxf->shapes()->color( draw.color );
+            dxf->shapes()->fill();
+            dxf->shapes()->polygen( v );
```
Comment (JamesBremner) [inline src/windex.cpp:301]: This does not compile.

`C:\Users\James\code\DXF_Viewer\src\windex.cpp|301|error: 'class wex::shapes' has no member named 'polygen'|`

**Comment by JamesBremner** [comment]
I think now that a better design would be to settle on a single internal representation of solids, no matter what DXF parsing option is used.  The benefit of this would be that the various drawing functions ( one for each gui framework ) would not need to coded for different parsing options.

For generality, I suppose that the internal representation should be a triangular mesh which can represent any solid - in particular a rectangle can be represented by two triangles.

I am slightly reluctant to implement this, because it makes the handling of the solids for my clients, who always use rectangles, more complex than they need.  But I do believe it is the correct way to go for a "general" DXF viewer.

File: src/windex.cpp (hunk 1 above)
Comment (asmwarrior) [reply src/windex.cpp:301]: > 
> 
> This does not compile.
> 
> `C:\Users\James\code\DXF_Viewer\src\windex.cpp|301|error: 'class wex::shapes' has no member named 'polygen'|`

Hi, please see this comment:
https://github.com/JamesBremner/DXF_Viewer/pull/25#issuecomment-599184798

you have to see my pull request to the windex project first.

Comment (JamesBremner) [reply src/windex.cpp:301]: Sorry. I missed notification of pull request.

**Comment by JamesBremner** [comment]
I would appreciate any comments you @asmwarrior  might have on this commit 2e1c342ccf73a6c529d6c607402248317a1f4a8e before I merge it into the master.

**Comment by asmwarrior** [comment]
> 
> 
> I would appreciate any comments you @asmwarrior might have on this commit [2e1c342](https://github.com/JamesBremner/DXF_Viewer/commit/2e1c342ccf73a6c529d6c607402248317a1f4a8e) before I merge it into the master.

I understand the logic you have made in the commit.

First, you add a new point(x3,y3) to the cDrawPrimitiveData.
Second, when you reading the two points of the rectangle, you just convert it to *two* triangles by the function call convert_2point().
Third, you draw the *two* triangles.

I'm OK with your changes.
Your code only support reading the 2 points (code=10 20, code=12 22).

And we can support the 4 points  (code=10 20, code = 11 21, code=12 22, code= 13 23) solid by adjust the code in bool CSolid::Append( cvit_t& cvit) to construct either one triangle or two triangles for the solid. ( From the document of dxf, I see the only support two kinds: one triangle kind of solid or two triangles kind of solid, which is the  quadrilateral)

Because you have a member `myTriangMesh` to remember all the triangles in the solid.
```
class CSolid  : public cDXFGraphObject
{
public:
    double x,y,x2,y2;       ///< locations of top-right and bottm-left corners
    std::vector< std::vector< std::vector< double > > > myTriangMesh
```

**Comment by JamesBremner** [comment]
> I'm OK with your changes.

Great.  Thanks for looking at the code.  I will merge it into the master.

> your code only support reading the 2 points (code=10 20, code=12 22).

Correct.  Also, I have not updated the wxWidgets draw function.  Please pull the merged master, implement these and create a new pull request.

**Comment by JamesBremner** [comment]
Merge successful.

Closing this pull request.

**Pull request closed without merging**